tasks/vagrant.py

```diff
@@ -5,6 +5,38 @@
 from .common import PopenTask, PopenException, FallibleTask, TaskException
 
 
+def with_vagrant(func):
+    def wrapper(self, *args, **kwargs):
+        try:
+            __setup_provision(self)
+        except TaskException as exc:
+            logging.critical('vagrant or provisioning failed')
+            raise exc
+        else:
+            func(self, *args, **kwargs)
+        finally:
+            if not self.no_destroy:
+                self.execute_subtask(
+                    VagrantCleanup(raise_on_err=False))
+
+    return wrapper
+
+
+def __setup_provision(self):
```

Comment:
I don't mind the `self` usage in the decorator's `wrapper` function, but I'd rather avoid it here, because it's a separate function and it's a bit confusing to have the `self` keyword outside a class. `task` might be better here.